### ISSUE TITLE
Merging imports deletes the "::*" suffix

### ISSUE THREAD

**Issue by nicomem** [rust-analyzer/rust-analyzer#3728]
## Problem description
Hi, I got the following problem:
when I'm trying to merge two imports where at least one is a "import-all", the "::*" suffix is deleted and thus breaks the import.

## Example
```rust
use std::cell::*;
use std::str;
```
*Merging*
```rust
use std::{cell, str};
```
It happens when either one (or both) is the "import-all"

## Version
- rust-analyser VS Code extension 0.2.100

**Comment by piotr-szpetkowski** [comment]
I'd like to work on it - unless it's the expected result in this case? @matklad

**Comment by matklad** [comment]
No, it is not an expected result. @piotr-szpetkowski my *guess* (which might be wrong) is that `ast::UseTree::split_prefix` function should be tweaked to translate `use std::cell::*` into `use std::cell::{*}`.

**Comment by piotr-szpetkowski** [comment]
@matklad Quick investigation tells that the problem might lay in the parsing. I've wrote a new test for `merge_imports` assist that verifies the result of merging with wildcard use (e.g. `use std::cell::*`) and the tree that I'm getting is something like this:

```
syntax: USE_TREE@[5; 17)
  PATH@[5; 14)
    PATH@[5; 8)
      PATH_SEGMENT@[5; 8)
        NAME_REF@[5; 8)
          IDENT@[5; 8) "std"
    COLONCOLON@[8; 10) "::"
    PATH_SEGMENT@[10; 14)
      NAME_REF@[10; 14)
        IDENT@[10; 14) "cell"
  COLONCOLON@[14; 16) "::"
  STAR@[16; 17) "*"
```

While a working one looks like this:

```
USE_TREE@[5; 20)
  PATH@[5; 20)
    PATH@[5; 14)
      PATH@[5; 8)
        PATH_SEGMENT@[5; 8)
          NAME_REF@[5; 8)
            IDENT@[5; 8) "std"
      COLONCOLON@[8; 10) "::"
      PATH_SEGMENT@[10; 14)
        NAME_REF@[10; 14)
          IDENT@[10; 14) "cell"
    COLONCOLON@[14; 16) "::"
    PATH_SEGMENT@[16; 20)
      NAME_REF@[16; 20)
        IDENT@[16; 20) "Cell"
```

Seems like the problem here is that the `STAR` node is not under a `PATH_SEGMENT` and therefore it's excluded out of the path during the `try_merge_trees` call.

There are two solutions that come to mind - first is to modify the structure of nodes to represent the `STAR` node (obviously only the one meeting the proper criteria) with a `PATH_SEGMENT` parent and the other solution might be to simply hack some ugly logic to make `STAR` part of the path, but I'm not sure if that's easily possible because the `STAR` node is out of the `PATH` tree and IMO it's not the responsibility of the path code to fix the tree issues.

**Comment by matklad** [comment]
I think this works as expected: in `use foo::{bar}` and `use foo::*`, the `::{bar}` and `::*` bits are delibireately not nested under paths, as they are not paths.

**Comment by piotr-szpetkowski** [comment]
Yes, my quick investigation was entirely wrong :slightly_smiling_face: Scratch that.

I have a basic understanding of the problem now and I think I might have an idea of how to fix it (which is actually close to your first reply here). I'll start working on a fix tomorrow and I believe I'll be able to submit a PR around the weekend.

Thanks for the explanations!

**Comment by matklad** [comment]
>I'll start working on a fix tomorrow and I believe I'll be able to submit a PR around the weekend.

Tip: when doing open source stuff on your free time, it's better not to name specific estimates, so that you don't feel pressured to actually abide by them :)

**Comment by eminence** [comment]
It appears that this issue can be closed, as this feature was implemented in #3745
